Libraries/LibCrypto/PK/RSA.cpp

```diff
@@ -0,0 +1,309 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibCrypto/ASN1/ASN1.h>
+#include <LibCrypto/ASN1/DER.h>
+#include <LibCrypto/ASN1/PEM.h>
+#include <LibCrypto/PK/RSA.h>
+
+namespace Crypto {
+namespace PK {
+
+RSA::KeyPairType RSA::parse_rsa_key(const ByteBuffer& in)
+{
+    // we are going to assign to at least one of these
+    KeyPairType keypair;
+    // TODO: move ASN parsing logic out
+    u64 t, x, y, z, tmp_oid[16];
+    u8 tmp_buf[4096] { 0 };
+    UnsignedBigInteger n, e, d;
+    ASN1::List pubkey_hash_oid[2], pubkey[2];
+
+    ASN1::set(pubkey_hash_oid[0], ASN1::Kind::ObjectIdentifier, tmp_oid, sizeof(tmp_oid) / sizeof(tmp_oid[0]));
+    ASN1::set(pubkey_hash_oid[1], ASN1::Kind::Null, nullptr, 0);
+
+    // DER is weird in that it stores pubkeys as bitstrings
+    // we must first extract that crap
+    ASN1::set(pubkey[0], ASN1::Kind::Sequence, &pubkey_hash_oid, 2);
+    ASN1::set(pubkey[1], ASN1::Kind::Null, nullptr, 0);
+
+    dbg() << "we were offered " << in.size() << " bytes of input";
+
+    if (der_decode_sequence(in.data(), in.size(), pubkey, 2)) {
+        // yay, now we have to reassemble the bitstring to a bytestring
+        t = 0;
+        y = 0;
+        z = 0;
+        x = 0;
+        for (; x < pubkey[1].size; ++x) {
+            y = (y << 1) | tmp_buf[x];
+            if (++z == 8) {
+                tmp_buf[t++] = (u8)y;
+                y = 0;
+                z = 0;
+            }
+        }
+        // now the buffer is correct (Sequence { Integer, Integer })
+        if (!der_decode_sequence_many<2>(tmp_buf, t,
+                ASN1::Kind::Integer, 1, &n,
+                ASN1::Kind::Integer, 1, &e)) {
+            // something was fucked up
+            dbg() << "bad pubkey: " << e << " in " << n;
+            return keypair;
+        }
+        // correct public key
+        keypair.public_key.set(n, e);
+        return keypair;
+    }
+
+    // could be a private key
+    if (!der_decode_sequence_many<1>(in.data(), in.size(),
+            ASN1::Kind::Integer, 1, &n)) {
+        // that's no key
+        // that's a death star
+        dbg() << "that's a death star";
+        return keypair;
+    }
+
+    if (n == 0) {
+        // it is a private key
+        UnsignedBigInteger zero;
+        if (!der_decode_sequence_many<4>(in.data(), in.size(),
+                ASN1::Kind::Integer, 1, &zero,
+                ASN1::Kind::Integer, 1, &n,
+                ASN1::Kind::Integer, 1, &e,
+                ASN1::Kind::Integer, 1, &d)) {
+            dbg() << "bad privkey " << n << " " << e << " " << d;
+            return keypair;
+        }
+        keypair.private_key.set(n, d, e);
+        return keypair;
+    }
+    if (n == 1) {
+        // multiprime key, we don't know how to deal with this
```

Comment:
everything else returns an empty keypair, I don't think asserting on such remote-controlled data is a good idea.

I am...unsure about this.

Reply:
Maybe `ASSERT`ing is too strong, yes.

Overall, the `parse_rsa_key` feels like it should return an `Optional<RSA::KeyPairType>` given that it fails so often, but the public and private keys are also "optional" (The `import_{*}_key` methods are even checking for existence of the individual keys with the `public_key.length()` and `private_key.length()` methods like they would with `.has_value()`).

I think maybe replacing the `KeyPairType` members with `Optional<Public/Private>` keys instead of direct values seems like it would be more explicit, but it would also be more unwieldy / harder to use so maybe the current code is fine too.